language-plutus-core/src/Language/PlutusCore/Check/ValueRestriction.hs

```diff
@@ -0,0 +1,32 @@
+-- | The value restriction check.
+
+{-# LANGUAGE LambdaCase #-}
+
+module Language.PlutusCore.Check.ValueRestriction
+    ( checkValueRestriction
+    , checkValueRestrictionProgram
+    ) where
+
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Type
+import           Language.PlutusCore.View  (isTermValue)
+
+import           Control.Monad.Error.Lens
+import           Control.Monad.Except
+import           Data.Foldable             (traverse_)
+import           Data.Functor.Foldable
+
+-- | Check whether a term satisfies the value restriction.
+checkValueRestriction
+    :: (AsValueRestrictionError e tyname ann, MonadError e m) => Term tyname name ann -> m ()
+checkValueRestriction = para $ \case
```

Comment:
Cute. The alternative would be to return a `m (Term Tyname Name ann)`, which has both the new body and the error effect. You can then discard it later. OTOH, since we don't want it this is neat. I'm not sure which I prefer.

Reply:
Sorry, I'm not following. What is the new body and why can it be useful?

Reply:
The difference between:
```
para f where f :: TermF tyname name a (Term tyname name a, m ()) -> m ()
```
and
```
cata f where f :: TermF tyname name a (m (Term tyname name a)) -> m (Term tyname name a)
```
Since you have no content in the result of your paramorphism, you could use it to carry the context. But I think it's just stylistic, really, and your version makes sense too.

Reply:
Ah, I see, thanks. I prefer the version with `para`. I'd also expect it to be more efficient.